Advanced Dataframes Exercise Problems

In [1]:
import pandas as pd
import numpy as np
from env import host, user, password, get_db_url

Exercises Part 1

Problem 1 - 3

Create a function named get_db_url. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.

In [19]:
def get_db_url(username, hostname, password, db_name):
    url = f'mysql+pymysql://{username}:{password}@{hostname}/{db_name}'
    return url

This is in the env.py file.

Problem 4: Use your function to obtain a connection to the employees database.

In [2]:
url = get_db_url(user, host, password, 'employees')

In [23]:
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', url)

emp_no  birth_date first_name  last_name gender   hire_date
0   10051  1953-07-28   Hidefumi      Caine      M  1992-10-15
1   10052  1961-02-26     Heping     Nitsch      M  1988-05-21
2   10053  1954-09-13     Sanjiv   Zschoche      F  1986-02-04
3   10054  1957-04-04     Mayumi  Schueller      M  1995-03-13
4   10055  1956-06-06     Georgy     Dredge      M  1992-04-27

Problem 5:

a. Intentionally make a typo in the database url. What kind of error message do you see?

In [33]:
url = get_db_url(user, host, password, 'employs')

In [27]:
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', url)

OperationalError: (pymysql.err.OperationalError) (1044, "Access denied for user 'innis_1658'@'%' to database 'employe'")
(Background on this error at: https://sqlalche.me/e/14/e3q8)

Returns an access denied error

b. Intentionally make an error in your SQL query. What does the error message look like?

In [34]:
url = get_db_url(user, host, password, 'employees')

In [29]:
pd.read_sql('SELECT * FROM employes LIMIT 5 OFFSET 50', url)

ProgrammingError: (pymysql.err.ProgrammingError) (1146, "Table 'employees.employes' doesn't exist")
[SQL: SELECT * FROM employes LIMIT 5 OFFSET 50]
(Background on this error at: https://sqlalche.me/e/14/f405)

It access the database but since i mispelled the employees talbe in the query it lets me know the table does not exist.

Problem 6: Read the employees and titles tables into two separate DataFrames.

In [4]:
employees = pd.read_sql('SELECT * FROM employees', url)
employees

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [5]:
titles = pd.read_sql('SELECT * FROM titles', url)
titles

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

Problem 7: How many rows and columns do you have in each DataFrame? Is that what you expected?

In [50]:
employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300024 entries, 0 to 300023
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   emp_no      300024 non-null  int64 
 1   birth_date  300024 non-null  object
 2   first_name  300024 non-null  object
 3   last_name   300024 non-null  object
 4   gender      300024 non-null  object
 5   hire_date   300024 non-null  object
dtypes: int64(1), object(5)
memory usage: 13.7+ MB


I have 300024 rows and 6 columns in the employees DataFrame.

In [51]:
titles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


I have 443308 rows and 4 columns in the titles DataFrame

Problem 8: Display the summary statistics for each DataFrame.

In [7]:
employees.describe()

emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

In [8]:
titles.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

Problem 9: How many unique titles are in the titles DataFrame?

In [12]:
len(pd.read_sql('SELECT title FROM titles GROUP BY title', url))

7

Problem 10: What is the oldest date in the to_date column?

In [13]:
pd.read_sql('SELECT to_date FROM titles', url).min()

to_date    1985-03-01
dtype: object

Problem 11: What is the most recent date in the to_date column?

In [18]:
pd.read_sql('SELECT to_date FROM titles WHERE to_date != "9999-01-01"', url).max()

to_date    2002-08-01
dtype: object

Exercise Part 2

Problem 1: What is the result of using a right join on the DataFrames?

Problem 2: What is the result of using an outer join on the DataFrames?

Problem 3: What happens if you drop the foreign keys from the DataFrames and try to merge them?

Problem 4: Load the mpg dataset from PyDataset.

Problem 5: Output and read the documentation for the mpg dataset.

Problem 6: How many rows and columns are in the dataset?

Problem 7: Check out your column names and perform any cleanup you may want on them.

Problem 8: Display the summary statistics for the dataset.

Problem 9: How many different manufacturers are there?

Problem 10: How many different models are there?

Problem 11: Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.

Problem 12: Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

Problem 13: Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission.

Problem 14: Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average?

Problem 15: Do automatic or manual cars have better miles per gallon?